useless-super-delegation when an override changes the return type annotation

### Current problem

Similar to #1923, but for the return type annotation.
It is sometimes useful to override a method and change the return type annotation - whether to help a static type checker out or to help the IDE make better autocomplete suggestions.

Here's an example - it's a bit contrived, I actually ran into this when overriding to add typing to a method from a third-party library (ANTLR).

```python
"""Quick test to demonstrate useless-super-declaration raised for overrides
that change return type annotations."""
import random
from dataclasses import dataclass
from typing import Any

# pylint: disable=too-few-public-methods
class _AGenerator:
    """Abstract base class for generators."""
    _valid_values: list[Any]

    def get_random(self) -> Any:
        """Generates a random value from the valid values that this generator
        can return."""
        return random.choice(self._valid_values)

@dataclass
class StrGenerator(_AGenerator):
    """Generator that accepts and returns string values."""
    _valid_values: list[str]

    # override to add typing
    def get_random(self) -> str:
        return super().get_random()

s = StrGenerator(['foo', 'bar', 'qux'])
print(s.get_random())
print(s.get_random())
print(s.get_random())
```

Running pylint with default config on it:
```
$ python -m pylint test.py 
************* Module test
test.py:24:4: W0235: Useless super delegation in method 'get_random' (useless-super-delegation)

------------------------------------------------------------------
Your code has been rated at 9.33/10 (previous run: 9.33/10, +0.00)
```

### Desired solution

pylint should not emit a warning in this case as the override has legitimate uses.

### Additional context

_No response_

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1299,8 +1299,8 @@ a metaclass class method.",
                     return
 
             if (
-                function.returns is not None and
-                meth_node.returns is None
+                function.returns is not None
+                and meth_node.returns is None
                 or meth_node.returns.as_string() != function.returns.as_string()
             ):
                 # Override adds typing information to the return type

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1299,9 +1299,9 @@ a metaclass class method.",
                     return
 
             if (
-                function.returns is not None and
-                meth_node.returns is not None and
-                meth_node.returns.as_string() != function.returns.as_string()
+                function.returns is not None
+                and meth_node.returns is not None
+                and meth_node.returns.as_string() != function.returns.as_string()
             ):
                 # Override adds typing information to the return type
                 return

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1299,9 +1299,9 @@ a metaclass class method.",
                     return
 
             if (
-                function.returns is not None
-                and meth_node.returns is None
-                or meth_node.returns.as_string() != function.returns.as_string()
+                function.returns is not None and
+                meth_node.returns is not None and
+                meth_node.returns.as_string() != function.returns.as_string()
             ):
                 # Override adds typing information to the return type
                 return

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1298,6 +1298,14 @@ a metaclass class method.",
                 if called_annotations != overridden_annotations:
                     return
 
+            if (
+                function.returns is not None and
+                meth_node.returns is None
+                or meth_node.returns.as_string() != function.returns.as_string()
+            ):
+                # Override adds typing information to the return type
+                return
+
         if _definition_equivalent_to_call(params, args):
             self.add_message(
                 "useless-super-delegation", node=function, args=(function.name,)

--- a/tests/functional/u/useless/useless_super_delegation_py35.py
+++ b/tests/functional/u/useless/useless_super_delegation_py35.py
@@ -36,7 +36,7 @@ class NoReturnType:
 class ReturnTypeSpecified(NoReturnType):
     choices = ['a', 'b']
 
-    def draw(self) -> str:
+    def draw(self) -> str: # [useless-super-delegation]
         return super().draw()
 
 class ReturnTypeSame(ReturnTypeAny):

--- a/tests/functional/u/useless/useless_super_delegation_py35.py
+++ b/tests/functional/u/useless/useless_super_delegation_py35.py
@@ -10,3 +10,37 @@ class UselessSuper(object):
 
     def useless(self, first, *, second=None, **kwargs): # [useless-super-delegation]
         return super().useless(first, second=second, **kwargs)
+
+# pylint: disable=wrong-import-position
+import random
+from typing import Any
+
+class ReturnTypeAny:
+    choices = ['a', 1, (2, 3)]
+
+    def draw(self) -> Any:
+        return random.choice(self.choices)
+
+class ReturnTypeNarrowed(ReturnTypeAny):
+    choices = [1, 2, 3]
+
+    def draw(self) -> int:
+        return super().draw()
+
+class NoReturnType:
+    choices = ['a', 1, (2, 3)]
+
+    def draw(self):
+        return random.choice(self.choices)
+
+class ReturnTypeSpecified(NoReturnType):
+    choices = ['a', 'b']
+
+    def draw(self) -> str:
+        return super().draw()
+
+class ReturnTypeSame(ReturnTypeAny):
+    choices = ['a', 'b']
+
+    def draw(self) -> Any: # [useless-super-delegation]
+        return super().draw()

--- a/tests/functional/u/useless/useless_super_delegation_py35.txt
+++ b/tests/functional/u/useless/useless_super_delegation_py35.txt
@@ -1,2 +1,3 @@
 useless-super-delegation:11:4:11:15:UselessSuper.useless:Useless super delegation in method 'useless':UNDEFINED
+useless-super-delegation:39:4:39:12:ReturnTypeSpecified.draw:Useless super delegation in method 'draw':UNDEFINED
 useless-super-delegation:45:4:45:12:ReturnTypeSame.draw:Useless super delegation in method 'draw':UNDEFINED

--- a/tests/functional/u/useless/useless_super_delegation_py35.txt
+++ b/tests/functional/u/useless/useless_super_delegation_py35.txt
@@ -1,1 +1,2 @@
 useless-super-delegation:11:4:11:15:UselessSuper.useless:Useless super delegation in method 'useless':UNDEFINED
+useless-super-delegation:45:4:45:12:ReturnTypeSame.draw:Useless super delegation in method 'draw':UNDEFINED

